### DISCUSSION TITLE
Help needed with writing and testing a function for Kotlin

### DISCUSSION THREAD

**Discussion by jborgers** [pmd/pmd#4957] [Q&A] [answered]
I created a function `hasImport` and now I want to test it with a test class like `TypeIsFunctionTest`.
In pmd-java module, I see folder `test/kotlin` and I added:

`net.sourceforge.pmd.lang.kotlin.rule.xpath.internal.HasImportFunctionTest `

there. Import dependencies are available there, so that is easy. However, I get the following error:
```[ERROR] net.sourceforge.pmd.lang.kotlin.rule.xpath.internal.HasImportFunctionTest.testFileHasNotFullImport  Time elapsed: 0.001 s  <<< ERROR!
net.sourceforge.pmd.lang.rule.xpath.PmdXPathException: 
net.sf.saxon.trans.XPathException: Namespace prefix 'pmd-kotlin' has not been declared
Exception Context:
	[1:Kind=Static error]
	[2:Phase=INITIALIZATION]
	[3:Expression=/KotlinFile[pmd-kotlin:hasImport('javax.xml.xpath')]]
	[4:Version=3.1]
	[5:Location in expr=; Line#: -1; Column#: 12]
	[6:Rule=$rule_name]
---------------------------------
	at net.sourceforge.pmd.lang.kotlin.rule.xpath.internal.HasImportFunctionTest.testFileHasNotFullImport(HasImportFunctionTest.java:52)
Caused by: net.sf.saxon.trans.XPathException: **Namespace prefix 'pmd-kotlin' has not been declared**
	at net.sourceforge.pmd.lang.kotlin.rule.xpath.internal.HasImportFunctionTest.testFileHasNotFullImport(HasImportFunctionTest.java:52)
```
Also, it seems Java syntax is applied for parsing the test code. 

I defined the namespace and function in xpath_funs.yml.

How to fix this?

Maybe the test class (including the dependencies) should go into pmd-kotlin module?

**Comment by jsotuyod** [comment, marked as answer]
`xpath_funs.yml` is just for documentation purposes. To get the function registered you would need to:

1. override `LanguageProcessor createProcessor(LanguagePropertyBundle bundle)` in `KotlinLanguagemodule` to provide your own processor extending `BatchLanguageProcessor` 
2. in your processor, override `XPathHandler getXPathHandler()` to return an `XPathHandler.getHandlerForFunctionDefs(YOUR_FUNCTION)`

You can see how this is done in Java here: https://github.com/pmd/pmd/blob/c472fdf48cdf1cb8827b0ff344ff1225f89487fe/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/JavaLanguageModule.java#L60-L63

And here: https://github.com/pmd/pmd/blob/c472fdf48cdf1cb8827b0ff344ff1225f89487fe/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/internal/JavaLanguageProcessor.java#L96-L99

https://github.com/pmd/pmd/blob/c472fdf48cdf1cb8827b0ff344ff1225f89487fe/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/internal/JavaLanguageProcessor.java#L116-L127

**Reply by jborgers** [reply]
Okay, thanks Juan Martin. I see much difference with the branch we are working on, need to fix that first.
